### PR TITLE
feat: generic webhook

### PR DESCRIPTION
**Pull request by arbreezy**
Closes #  #133 

## ✅ Checks

- [ ] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [ ] All the tests have passed

## ℹ Additional Information

I have created a few Results functions in a similar fashion with k8sgpt resources, introduced two status fields for Results 

1. Lifecycle -> to show the lifecycle of a Result CR {created, updated, historical}
2. Webhook -> to show the sink endpoint that is configured in K8sGPT CR for each Result 

In regard to emitting messages I depend on the status of each Result to send its payload to the specified sink while also updating/cleaning up the status if needed

Sink is an interface that has two signatures Configure and Emit which may be useful if we want to add more sink types ( discord, etc..)

We also cleanup the statuses of Results when Sink is not defined anymore in K8sGPT CR

PS: I didn't consider a fanout solution cause I haven't seen such need to configure more than one integrations.

### CODE REVIEW

**Comment by AnaisUrlichs** [comment]
Could you please update the following page in the docs with the minimum information needed for people to 
1. know that there is a webhook option
2. understand how to make use of it
Link to docs:
https://docs.k8sgpt.ai/reference/operator/overview/#customising-the-operator

File: api/v1alpha1/k8sgpt_types.go [hunk 1]
```diff
@@ -36,11 +36,16 @@ type ExtraOptionsRef struct {
 	Backstage *Backstage `json:"backstage,omitempty"`
 }
 
+type WebhookRef struct {
+	Type     string `json:"type,omitempty"`
```
Comment (AlexsJones) [inline api/v1alpha1/k8sgpt_types.go:40]: We should use the validation comment to ensure this is only set to slack

File: pkg/sinks/slack.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+func buildSlackMessage(kind, name, details, backend string) SlackMessage {
+	return SlackMessage{
+		Text: fmt.Sprintf("`Analysis from %s of the %s %s`", backend, kind, name),
+		Attachments: []Attachment{
+			Attachment{
+				Type:  "mrkdwn",
+				Text:  details,
+				Color: "danger",
```
Comment (AlexsJones) [inline pkg/sinks/slack.go:38]: I anticipate eventually this will need to be configurable somehow

File: pkg/sinks/slack.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+func (s *SlackSink) Configure(config v1alpha1.K8sGPT, c Client) {
+	if config.Spec.Sink == nil {
+		s.Endpoint = ""
+	}
+	s.Endpoint = config.Spec.Sink.Endpoint
+	s.Client = c
+}
+
+func (s *SlackSink) Emit(results v1alpha1.ResultSpec) error {
+	if s.Endpoint == "" {
```
Comment (AlexsJones) [inline pkg/sinks/slack.go:54]: There should be some sort of error that returns a warning otherwise this will be very frustrating

**Review by AlexsJones — CHANGES_REQUESTED** [review]
Left some messages inline

File: pkg/sinks/slack.go (hunk 2 above)
Comment (arbreezy) [reply pkg/sinks/slack.go:38]: which parts are you addressing to be configurable from the user?

File: pkg/sinks/slack.go (hunk 3 above)
Comment (arbreezy) [reply pkg/sinks/slack.go:54]: My thought so far was that an empty endpoint comes from two different places,
a) Top level Sink struct is omitted so a user don't want to configure a webhook endpoint thus it's empty
b) a user sets an empty webhook in their CR ( not sure why to set a blank CR value) , also in that case endpoint is empty so we don't emit anything,

I think a message or warning is not easy with this implementation cause we don't know which of the two cases is real, 
finally, I am not sure how a user can be frustrated if they don't set any webhook url while expecting to integrate k8sgpt, but eager to get more feedback

Comment (arbreezy) [reply pkg/sinks/slack.go:54]: I also wanted to avoid more if conditions but looks more clunky, I will update the PR

**Comment by arbreezy** [comment]
Slack message looks like this,
Screenshot from 2023-06-08 17-23-48

File: pkg/sinks/slack.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+				Type:  "mrkdwn",
+				Text:  details,
+				Color: "danger",
+				Title: "Report",
+			},
+		},
+	}
+}
+
+func (s *SlackSink) Configure(config v1alpha1.K8sGPT, c Client) {
```
Comment (AlexsJones) [inline pkg/sinks/slack.go:45]: Why don't we pass this in by pointer? `config v1alpha1.K8sGPT`

File: pkg/sinks/slack.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+	s.Endpoint = config.Spec.Sink.Endpoint
+	s.Client = c
+}
+
+func (s *SlackSink) Emit(results v1alpha1.ResultSpec) error {
```
Comment (AlexsJones) [inline pkg/sinks/slack.go:50]: Also, why copy and not use a pointer?

**Review by AlexsJones — APPROVED** [review]

**Pull request merged**